### PR TITLE
[Quantization][FP8] Add support for FP8 models with input_scale for output projection and QK quantization

### PR DESCRIPTION
**Pull request by rasmith**
This PR adds fp8 quantization support for:

1) Preserving FP8 quantization after FA output, so that output of FA into the next layer will be FP8
     - this uses the model parameter self_attn.q/k/v_proj.input_scale and passes it as _out_scale to the FA kernel
2) During execution of FA loop, the quantity softmax(QK^T) can also be quantized as FP8
    - this uses the model parameter self_attn.prob_output_scale

These are using Quark quantized models, in particular using Llama-3.1-8B-Instruct-FP8-QKV-Prob

So far, only Llama is being supported, but more can be added later.

### CODE REVIEW

**Review by ProExpertProg — COMMENTED** [review]
As I think we've mentioned before, we should not be editing model files to manually fuse operations across layers of abstraction. Let's instead merge this PR without the fused output scaling and instead fuse using a `torch.compile` pass.

File: vllm/model_executor/layers/quantization/kv_cache.py [hunk 1]
```diff
@@ -97,5 +101,36 @@ def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
                     "may cause accuracy issues. Please make sure k/v_scale "
                     "scaling factors are available in the fp8 checkpoint.")
 
+        if layer.q_scale > 0.0:
+            q_scale = layer.q_scale.to("cpu").tolist()
+            if current_platform.is_rocm():
+                q_scale *= 2
```
[the remark this answers is not in the document]
Comment (rasmith) [reply vllm/model_executor/layers/quantization/kv_cache.py:107]: Changed to use is_fp8_fnuz() and keeping scales on device (not sure why the cpu thing was there).

File: vllm/attention/backends/rocm_flash_attn.py [hunk 2]
```diff
@@ -699,6 +699,13 @@ def forward(
                             query.dtype,
                             seq_lens,
                             make_attn_mask=causal_mask)  # type: ignore
+                    use_fp8_out_scale = (layer._out_scale and layer._q_scale
+                                         and layer._prob_scale
+                                         and self.kv_cache_dtype == "fp8")
+                    full_scales = (
+                        layer._q_scale.item(), layer._k_scale.item(),
+                        layer._v_scale.item(), layer._prob_scale.item(),
+                        layer._out_scale.item()) if use_fp8_out_scale else None
```
[the remark this answers is not in the document]
Comment (rasmith) [reply vllm/attention/backends/rocm_flash_attn.py:708]: The kernel takes a single floating-point value instead of a tensor.  If a tensor is passed, then incompatible type error will be generated.

File: vllm/model_executor/layers/quantization/utils/w8a8_utils.py [hunk 3]
```diff
@@ -216,7 +218,11 @@ def apply(
                     num_token_padding=self.output_padding,
                     use_per_token_if_dynamic=use_per_token_if_dynamic)
             else:
-                qinput, x_scale = input_2d, input_scale
+                assert (
+                    input_scale is not None and input_scale.shape.numel() == 1
+                ), "Only static per-tensor scaling implemented in this case."
+                qinput = triton_fp8_inplace_scale(input_2d, input_scale.item())
+                x_scale = input_scale
```
Comment (ProExpertProg) [inline vllm/model_executor/layers/quantization/utils/w8a8_utils.py:225]: As discussed in Slack, the path above should be taken here

File: vllm/model_executor/layers/quantization/kernels/fp8_inplace_scale.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
+# SPDX-License-Identifier: Apache-2.0
```
Comment (ProExpertProg) [inline vllm/model_executor/layers/quantization/kernels/fp8_inplace_scale.py:1]: We should be able to remove this

File: tests/kernels/test_triton_fp8_inplace_scale.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
+# SPDX-License-Identifier: Apache-2.0
```
Comment (ProExpertProg) [inline tests/kernels/test_triton_fp8_inplace_scale.py:1]: We should be able to remove this

File: vllm/attention/backends/rocm_flash_attn.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ProExpertProg) [reply vllm/attention/backends/rocm_flash_attn.py]: Could we change the kernel to take the tensor? I think usually for scales we just pass tensors around even if it's a single value.

[the remark this answers is not in the document]
Comment (rasmith) [reply vllm/attention/backends/rocm_flash_attn.py]: Yes, but the kernel still only supports a single fp32 value, which means the .item() will still be called, just in the wrapper code.  Unless you want every single fp32 value here to be passed as a pointer to the kernel and then loaded via tl.load().

File: vllm/model_executor/layers/quantization/utils/w8a8_utils.py (hunk 3 above)
Comment (rasmith) [reply vllm/model_executor/layers/quantization/utils/w8a8_utils.py]: Switched it back

File: vllm/attention/backends/rocm_flash_attn.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ProExpertProg) [reply vllm/attention/backends/rocm_flash_attn.py]: Yeah let's pass the pointers and load explicitly. Otherwise we're just moving the values between GPU and CPU unnecessarily, right?

File: vllm/model_executor/layers/quantization/kernels/fp8_inplace_scale.py (hunk 4 above)
Comment (rasmith) [reply vllm/model_executor/layers/quantization/kernels/fp8_inplace_scale.py]: Removed

File: tests/kernels/test_triton_fp8_inplace_scale.py (hunk 5 above)
Comment (rasmith) [reply tests/kernels/test_triton_fp8_inplace_scale.py]: Removed

File: vllm/model_executor/layers/quantization/fp8.py [hunk 6]
```diff
@@ -140,6 +140,11 @@ def get_cache_scale(self, name: str) -> Optional[str]:
             return name.replace(".k_proj.output_scale", ".attn.k_scale")
         if name.endswith(".output_scale") and ".v_proj" in name:
             return name.replace(".v_proj.output_scale", ".attn.v_scale")
+        if name.endswith(".output_scale") and ".q_proj" in name:
```
Comment (BowenBao) [inline vllm/model_executor/layers/quantization/fp8.py:143]: Is it necessary to extend this part if already handled in `quark.py`? cc @kewang-xlnx 

Same question applies for change to `compressed_tensors.py` above.

Comment (rasmith) [reply vllm/model_executor/layers/quantization/fp8.py:143]: It was necessary, I had to add this for it to work.

Comment (rasmith) [reply vllm/model_executor/layers/quantization/fp8.py:143]: OK, just checked again, the compressed tensors modification is not necesssary, unless maybe there is int8 w8a8 output_scale.  But some FP8 models will be quantized as "fp8" in quantization config, and some are just "quark", so the renaming needs to happen in both places.

File: vllm/model_executor/layers/quantization/kv_cache.py [hunk 7]
```diff
@@ -97,5 +100,35 @@ def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
...
 
+        if layer.q_scale > 0.0:
+            q_scale = layer.q_scale.tolist()
+            if current_platform.is_fp8_fnuz():
+                q_scale *= 2
+            layer.calculate_kv_scales = False
+        else:
+            q_scale = 1.0
+        if layer.prob_scale > 0.0:
+            prob_scale = layer.prob_scale.to("cpu").tolist()
```
Comment (ProExpertProg) [inline vllm/model_executor/layers/quantization/kv_cache.py:111]: Still going to CPU unnecessarily

File: vllm/model_executor/layers/quantization/kv_cache.py [hunk 8]
```diff
@@ -97,5 +101,36 @@ def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
...
+
+        if not isinstance(q_scale, float) or not isinstance(prob_scale, float):
+            raise ValueError("Only support per-tensor scaling factor"
+                             "for fp8-quantized Q/prob")
+
+        # These are used in the final Attention.forward()
+        layer._q_scale.copy_(q_scale)
+        layer._prob_scale.copy_(prob_scale)
+        if (q_scale == 1.0
+                or prob_scale == 1.0) and envs.VLLM_USE_ROCM_FP8_FLASH_ATTN:
```
[the remark this answers is not in the document]
Comment (rasmith) [reply vllm/model_executor/layers/quantization/kv_cache.py]: Not anymore, I think somebody asked for its deletion.

File: vllm/model_executor/layers/quantization/kv_cache.py [hunk 9]
```diff
@@ -97,5 +101,36 @@ def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
...
+        if layer.q_scale > 0.0:
+            q_scale = layer.q_scale.tolist()
+            if current_platform.is_fp8_fnuz():
+                q_scale *= 2
+            layer.calculate_kv_scales = False
+        else:
+            q_scale = 1.0
+        if layer.prob_scale > 0.0:
+            prob_scale = layer.prob_scale.to("cpu").tolist()
+            if current_platform.is_rocm():
```
[the remark this answers is not in the document]
Comment (rasmith) [reply vllm/model_executor/layers/quantization/kv_cache.py:112]: Done

File: vllm/model_executor/layers/quantization/utils/w8a8_utils.py [hunk 10]
```diff
@@ -215,12 +215,14 @@ def apply(
                     input_scale,
                     num_token_padding=self.output_padding,
                     use_per_token_if_dynamic=use_per_token_if_dynamic)
+
+                per_tensor_weights = (weight_scale.numel() == 1)
+                per_tensor_activations = (x_scale.numel() == 1)
```
Comment (ProExpertProg) [inline vllm/model_executor/layers/quantization/utils/w8a8_utils.py:220]: This seems to be duplicated below?

**Review by ProExpertProg — COMMENTED** [review]
I think the whole scale logic (and going to CPU) is messy in the existing code. I'll create an issue that you can link with a TODO so that we can go back and address this

File: vllm/model_executor/layers/quantization/kv_cache.py (hunk 7 above)
Comment (rasmith) [reply vllm/model_executor/layers/quantization/kv_cache.py]: Stays on GPU now

File: vllm/model_executor/layers/quantization/utils/w8a8_utils.py (hunk 10 above)
Comment (rasmith) [reply vllm/model_executor/layers/quantization/utils/w8a8_utils.py]: De-duped

File: vllm/model_executor/layers/quantization/fbgemm_fp8.py [hunk 11]
```diff
@@ -36,6 +36,7 @@ def __init__(self, ignore_list: List[str], input_scale_ub: float):
         # kernel for fast weight-only FP8 quantization
         self.use_marlin = not current_platform.has_device_capability(89)
         self.fp8_linear = Fp8LinearOp()
+        self.out_dtype = torch.get_default_dtype()
```
Comment (SageMoore) [inline vllm/model_executor/layers/quantization/fbgemm_fp8.py:39]: Nit: Same idea here, we aren't accessing this now that the model definition changes have been moved out. Let's add this back in if we need it in a follow on PR.

File: vllm/model_executor/layers/quantization/kv_cache.py [hunk 12]
```diff
@@ -38,6 +38,9 @@ def create_weights(self, layer: torch.nn.Module):
                                            requires_grad=False)
         layer.v_scale = torch.nn.Parameter(torch.tensor(-1.0),
                                            requires_grad=False)
+        # Initialize Q and P = softmax(QK^T) scales
```
Comment (SageMoore) [inline vllm/model_executor/layers/quantization/kv_cache.py:41]: Nit: update comment to just "Initialize P ...."

**Review by SageMoore — COMMENTED** [review]
Looks reasonable @rasmith. I only have small nits.

File: vllm/model_executor/layers/quantization/fbgemm_fp8.py (hunk 11 above)
Comment (rasmith) [reply vllm/model_executor/layers/quantization/fbgemm_fp8.py:39]: Done

File: vllm/model_executor/layers/quantization/quark/quark.py [hunk 13]
```diff
@@ -151,6 +156,19 @@ def _check_scheme_supported(self,
         else:
             return False
 
+    def is_fp8_w8a8(self) -> bool:
```
[the remark this answers is not in the document]
Comment (rasmith) [reply vllm/model_executor/layers/quantization/quark/quark.py:159]: Done

**Review by SageMoore — APPROVED** [review]

File: vllm/attention/backends/abstract.py [hunk 14]
```diff
@@ -237,6 +237,7 @@ class AttentionLayer(Protocol):
     _v_scale: torch.Tensor
     _k_scale_float: float
     _v_scale_float: float
+    _prob_scale: torch.Tensor
```
Comment (tlrmchlsmth) [inline vllm/attention/backends/abstract.py:240]: QQ: why is this called `_prob_scale`?

**Review by tlrmchlsmth — COMMENTED** [review]
Does this only work with vLLM V0? Ideally this would get added to V1 (and then it wouldn't necessarily need to happen in V0 since we plan to deprecate it "soon").

What will happen if someone tries to run such a model in V1? 

BTW is `amd/Llama-3.1-8B-Instruct-FP8-KV` the right huggingface stub to try this out?

File: vllm/model_executor/layers/quantization/kv_cache.py [hunk 15]
```diff
@@ -97,5 +100,38 @@ def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
...
+            q_scale = 1.0
+        if layer.prob_scale > 0.0:
+            prob_scale = layer.prob_scale
+            if current_platform.is_fp8_fnuz():
+                prob_scale *= 2
+        else:
+            prob_scale = 1.0
+
+        is_singleton_float = lambda x: isinstance(x, float) or isinstance(
+            x, torch.Tensor) and x.numel() == 1
```
Comment (ProExpertProg) [inline vllm/model_executor/layers/quantization/kv_cache.py:118]:
```suggestion
        is_singleton_float = lambda x: isinstance(x, float) or isinstance(
            x, torch.Tensor) and x.numel() == 1 and x.dtype == torch.float32
```

File: vllm/attention/backends/abstract.py (hunk 14 above)
Comment (ProExpertProg) [reply vllm/attention/backends/abstract.py:240]: I think this is the scale for P, which is the tensor resultng from `softmax(Q@K)` calculation. `O` is `softmax(Q@K)@V)`

**Review by ProExpertProg — APPROVED** [review]

**Review by robertgshaw2-redhat — APPROVED** [review]

**Review by tlrmchlsmth — CHANGES_REQUESTED** [review]
Let’s make sure this doesn’t break V1 before landing

File: vllm/attention/backends/abstract.py (hunk 14 above)
Comment (rasmith) [reply vllm/attention/backends/abstract.py:240]: It comes from the parameter in the model, something like `self_attn.prob_output_scale`, it gets remapped to `.attn.prob_scale` and also @ProExpertProg 's comments.

**Comment by rasmith** [comment]
> Does this only work with vLLM V0? Ideally this would get added to V1 (and then it wouldn't necessarily need to happen in V0 since we plan to deprecate it "soon").
> 
> What will happen if someone tries to run such a model in V1?
> 
> BTW is `amd/Llama-3.1-8B-Instruct-FP8-KV` the right huggingface stub to try this out?

It's not V0 ready, but I think it was agreed that a follow-on PR would be OK.  You can use amd/Llama-3.1-8B-Instruct-FP8-KV or amd/Llama-3.1-70B-Instruct-FP8-QKV-Prob.  Or if the model has 'k_proj/v_proj/k_proj.output_scale' and is set for 'fp8' or 'quark' quantization.

File: vllm/model_executor/layers/quantization/kv_cache.py (hunk 15 above)
Comment (rasmith) [reply vllm/model_executor/layers/quantization/kv_cache.py:118]: Went with adding x.is_floating_point()

**Comment by rasmith** [comment]
> Let’s make sure this doesn’t break V1 before landing

OK, I added a check in arg_utils to check if fp8 quantization being used and it just calls _raise_or_fallback.

**Comment by rasmith** [comment]
@tlrmchlsmth Please take another look? Ran into some problem with VllmConfig._get_quantization_config, it modifies the model_config object that gets passed to it.

However, when running with VLLM_USE_V1=1 there is a check in arg_utils.py that will check for compatibility.  The check works for both cases of V0 and V1 now.  So, if using V0, everything works, and if  using VLLM_USE_V1=1 the check happens and user is just told that fp8/quark fp8 is not implemented for ROCm.

**Review by tlrmchlsmth — APPROVED** [review]
